<a href="https://colab.research.google.com/github/mrsferret/Code-Division/blob/main/Worksheets/10_2_Movies_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Movies Mini-project
---

In the previous worksheet you converted an SQL relational database to a single pandas dataframe and downloaded it. You will be analysing it today.

If you were unable to download the file, there is a copy located here: 

"https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"



### Inspect the dataset 
---

In [65]:
import pandas as pd

def inspect_data(df):

  # firstly, sort the data on movie, director and year
  df = df.sort_values(by = ["name", "last_name", "first_name", "year"])
  display(df.describe())
  display(df.head())
  display(df.info())

  # check the data for NA/NaN/missing values across the dataframe
  display("Null Values: ", df.isnull().values.any())

  #============================================================================
  # null values do exist in the dataset and df.info() indicates that column "rank"
  # is only column with them
  #============================================================================
  display("Null Values: ", df.isnull().values.any())
  '''' is any missing values in dataframe '''
  display(df.isnull())

  #============================================================================
  # lets look more closely at the rank column as we know it contains null values
  # check what unique values are stored in this column and its count
  #============================================================================
  print("\nUnique values from column rank...\n",df['rank'].unique())
  print("\nCount unique values from a column ...\n",df['rank'].nunique())

  ### get count of missing values of a particular column
 
  #display ("Count of missing values in rank col: ", df.rank.isnull().sum())
  #display (df.rank.isnull().sum())

  display("count of missing values: ", df.isnull().sum())
  display (df.isnull().sum()/len(df)*100)

  #============================================================================
  # lets look more closely at the genre column to check what values are stored 
  # in this column and its count
  #============================================================================

  print("\nUnique values from column genre...\n",df['genre'].unique())
  print("\nCount unique values from a column ...\n",df['genre'].nunique())

  #========================================================================
  # should only be 1 row per movie so check for duplicates
  #========================================================================
  duplicates = find_duplicates(df)
  #duplicates = df[df.duplicated(['name','last_name','year'])]
  #print("Duplicate Values")
  #print(duplicates)
  #print("\nCount unique values from name column ...\n",duplicates['name'].nunique())

  # check if all duplicates have missing values
  #display("count of missing values in duplicates: ", duplicates.isnull().sum())
  #display (duplicates.isnull().sum()/len(duplicates)*100)


def find_duplicates(df):
  #duplicates = df[df.duplicated(['name','last_name','year'])]
  duplicates = df[df.duplicated(['name'])]
  display("Duplicate Values")
  display(duplicates.sort_values(by = ["name", "last_name", "year"]))
  display("\nCount unique values from name column ...\n",duplicates['name'].nunique())

  # check if all duplicates have missing values
  display("count of missing values in duplicates: ", duplicates.isnull().sum())
  display (duplicates.isnull().sum()/len(duplicates)*100)

  return duplicates 
  

# program starts here
url = "https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"

movie_df = pd.read_csv(url)

inspect_data(movie_df)

Unnamed: 0           year           rank
count  368894.000000  368894.000000  113376.000000
mean   184446.500000    1969.442222       5.995124
std    106490.669438      29.082051       1.540330
min         0.000000    1888.000000       1.000000
25%     92223.250000    1947.000000       5.100000
50%    184446.500000    1976.000000       6.200000
75%    276669.750000    1996.000000       7.100000
max    368893.000000    2007.000000       9.900000

Unnamed: 0 first_name    last_name  \
11330        11330    Mohamed         Asli   
265567      265567      Diane      Pierens   
265568      265568      Diane      Pierens   
265569      265569      Diane      Pierens   
274883      274883  Francisco  Ramalho Jr.   

                                       name  year  rank   genre  
11330    Casablanca les anges ne volent pas  2004   NaN   Drama  
265567                                Clara  1994   NaN   Drama  
265568                                Clara  1994   NaN  Family  
265569                                Clara  1994   NaN   Short  
274883                         Flor da Pele  1976   5.5   Drama

<class 'pandas.core.frame.DataFrame'>
Int64Index: 368894 entries, 11330 to 218388
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   Unnamed: 0  368894 non-null  int64  
 1   first_name  368894 non-null  object 
 2   last_name   368894 non-null  object 
 3   name        368894 non-null  object 
 4   year        368894 non-null  int64  
 5   rank        113376 non-null  float64
 6   genre       368894 non-null  object 
dtypes: float64(1), int64(2), object(4)
memory usage: 22.5+ MB


None

'Null Values: '

True

'Null Values: '

True

Unnamed: 0  first_name  last_name   name   year   rank  genre
11330        False       False      False  False  False   True  False
265567       False       False      False  False  False   True  False
265568       False       False      False  False  False   True  False
265569       False       False      False  False  False   True  False
274883       False       False      False  False  False  False  False
...            ...         ...        ...    ...    ...    ...    ...
170647       False       False      False  False  False   True  False
170648       False       False      False  False  False   True  False
218386       False       False      False  False  False   True  False
218387       False       False      False  False  False   True  False
218388       False       False      False  False  False   True  False

[368894 rows x 7 columns]


Unique values from column rank...
 [nan 5.5 8.7 5.1 5.2 7.8 6.  7.2 4.  6.8 3.1 3.9 6.2 5.8 4.3 6.4 7.3 8.1
 7.5 8.2 3.2 7.9 6.3 5.6 7.1 5.  7.  6.7 6.5 9.6 5.4 6.6 4.9 7.7 6.1 3.5
 6.9 5.7 1.6 4.6 2.  4.8 5.9 3.8 7.4 2.4 4.4 4.7 8.  3.6 5.3 2.8 4.5 2.3
 4.1 8.8 3.3 2.2 9.  7.6 9.8 8.4 9.3 4.2 9.5 8.3 1.4 3.4 2.5 3.7 1.8 9.2
 3.  8.5 2.7 2.9 9.4 1.9 8.9 8.6 9.1 9.7 2.6 1.7 1.  2.1 1.5 1.3 9.9 1.1
 1.2]

Count unique values from a column ...
 90


'count of missing values: '

Unnamed: 0         0
first_name         0
last_name          0
name               0
year               0
rank          255518
genre              0
dtype: int64

Unnamed: 0     0.000000
first_name     0.000000
last_name      0.000000
name           0.000000
year           0.000000
rank          69.265968
genre          0.000000
dtype: float64


Unique values from column genre...
 ['Drama' 'Family' 'Short' 'Comedy' 'Documentary' 'Musical' 'Romance'
 'Fantasy' 'Crime' 'Horror' 'Mystery' 'War' 'Thriller' 'Animation'
 'Western' 'Music' 'Action' 'Sci-Fi' 'Adventure' 'Film-Noir']

Count unique values from a column ...
 20


'Duplicate Values'

Unnamed: 0    first_name last_name                            name  \
265568      265568         Diane   Pierens                           Clara   
265569      265569         Diane   Pierens                           Clara   
128111      128111  Pierre-Marie    Goulet                           Gaule   
45580        45580         Pedro    Caldas                  S Um Minuto...   
8947          8947        Hubert     Aquin   Saint-Henri le cinq septembre   
...            ...           ...       ...                             ...   
326242      326242        András   Szirtes                zenetek a Holdrl   
9113          9113         Aydin    Arakon                          zleyis   
170648      170648         Kemal       Kan                zmc kizin kaderi   
218387      218387     Seiji (I)  Maruyama                 zora no samurai   
218388      218388     Seiji (I)  Maruyama                 zora no samurai   

        year  rank    genre  
265568  1994   NaN   Family  
265569  1994   NaN    Short  
128111  1974   NaN    Short  
45580   1998   NaN    Short  
8947    1962   NaN    Short  
...      ...   ...      ...  
326242  1990   NaN    Short  
9113    1961   NaN  Romance  
170648  1961   NaN  Romance  
218387  1976   NaN    Drama  
218388  1976   NaN      War  

[171695 rows x 7 columns]

'\nCount unique values from name column ...\n'

99029

'count of missing values in duplicates: '

Unnamed: 0         0
first_name         0
last_name          0
name               0
year               0
rank          111607
genre              0
dtype: int64

Unnamed: 0     0.000000
first_name     0.000000
last_name      0.000000
name           0.000000
year           0.000000
rank          65.003058
genre          0.000000
dtype: float64



---


# **Inspecting the Data Notes**

---


1.   On initial look at the data, I can see that the dataset has **368894** rows. From  **.info()**, the only column that contains null data is the "**rank**" column. This will need to be cleaned. The first column in the dataframe is called "**Unnamed: 0**" and on initial inspection it looks like it is just a record number. Will need to look at this further.

2.   **Rank Column** - Looking more closely at the rank column which contains null values, we can see that there are **255518** records with a missing rank value and this represents nearly **70%** of the total number of recs. There are **90** unique values in this column (including nan) with a **max value of 9.9** and a **minimum of 1**. This leaves us with the question of what to do with the rank column? Options are to **a)** drop the column entirely from the dataframe **b)** somehow impute a sensible value for the empty fields **c)** remove only those records that have an empty rank value, although this would result in a dataset **reduced by nearly 70%**, a not insignificant amount.

1.  **Duplicate row check** - when checking for duplicates, no Movie/Director/Year combination should have more than 1 row. When check for rows with duplicate values, **138290** rows are returned (includes all duplicate rows) which represents **89140** rows with a unique movie name. From this initial check, it looks possible that all duplicate rows have missing values in the **rank** column. Check this out.
2.   After checking whether all duplicates have missing values in the **rank** column, I can see that this is not actually the case. Of **138290** duplicate rows returned, only **88465** of these have missing values in the rank field


2.   List item



 






### Clean the dataset 
---


In [69]:
from os import name
import pandas as pd

def clean_data(df):

  # ================================================================
  # remove all rows where column rank is null
  # ================================================================
  clean_df = df[~df["rank"].isnull()]
  display (clean_df)

  # ================================================================
  # remove leading/trailing spaces from strings
  # ================================================================
 
  # removing whitespace from more than 1 column
  clean_df['name'].str.strip()
  clean_df['first_name'].str.strip()
  clean_df['last_name'].str.strip()

  # dataframe
  #print("Dataframe after removing whitespaces...\n",clean_df)

  # ================================================================
  # now that non-nulls removed, check duplicate situation
  # ================================================================

  find_duplicates1(clean_df)
  display (clean_df)
  
  # selecting rows based on condition
  rslt_df = clean_df[clean_df['last_name'] == "Aaltonen"]
  num_rows = rslt_df.count()[0]

  print('Number of Rows in Aaltonen DataFrame :',num_rows)
 
  print('\nResult dataframe :\n', rslt_df)
  

  return clean_df

def find_duplicates1(df):
  #duplicates = df[df.duplicated(['name','last_name','year'])]
  duplicates = df[df.duplicated(['name'])]

  display("Duplicate Values")
  sort_df = duplicates.sort_values(by = ["name"])
  display ("Sorted: ", sort_df) 
  display("\nCount unique values from name column ...\n",sort_df['name'].nunique())

  # check if all duplicates have missing values
  display("count of missing values in duplicates: ", duplicates.isnull().sum())
  display (duplicates.isnull().sum()/len(duplicates)*100)

  # selecting rows based on condition
  rslt_df = df[df['last_name'] == "Aaltonen"]
  num_rows = rslt_df.count()[0]

  print('Number of Rows in Aaltonen DataFrame :',num_rows)
 
  print('\nResult dataframe :\n', rslt_df)

  return duplicates 

# program starts here
url = "https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"

movie_df = pd.read_csv(url)

clean_data(movie_df)



Unnamed: 0 first_name last_name                                name  \
77              77  Rostislav     Aalto                        Cleaning Up!   
78              78     Veikko  Aaltonen                            Is meidn   
79              79     Veikko  Aaltonen                     Juoksuhaudantie   
80              80     Veikko  Aaltonen                     Juoksuhaudantie   
81              81     Veikko  Aaltonen                                 Maa   
...            ...        ...       ...                                 ...   
368837      368837     Memduh        Ün                               Garip   
368838      368838     Memduh        Ün                               Garip   
368861      368861     Memduh        Ün              Sezercik aslan parcasi   
368862      368862     Memduh        Ün              Sezercik aslan parcasi   
368891      368891       Idil      Üner  Liebenden vom Hotel von Osman, Die   

        year  rank        genre  
77      2002   8.3  Documentary  
78      1993   6.8        Drama  
79      2004   6.3       Comedy  
80      2004   6.3        Drama  
81      2001   6.0  Documentary  
...      ...   ...          ...  
368837  1986   7.9       Comedy  
368838  1986   7.9        Drama  
368861  1972   6.0        Drama  
368862  1972   6.0      Romance  
368891  2001   8.5        Short  

[113376 rows x 7 columns]

'Duplicate Values'

'Sorted: '

Unnamed: 0   first_name last_name                    name  year  rank  \
233670      233670     Giuseppe    Moccia   arrivato mio fratello  1985   5.1   
216153      216153        Mario    Marret         bientt, j'espre  1968   5.2   
124273      124273     Jean-Luc    Godard         bout de souffle  1960   7.8   
124272      124272     Jean-Luc    Godard         bout de souffle  1960   7.8   
153855      153855    Jean-Loup    Hubert            cause d'elle  1993   6.0   
...            ...          ...       ...                     ...   ...   ...   
49586        49586         Marc      Caro              vasion, L'  1978   7.9   
165031      165031  Jean-Pierre    Jeunet              vasion, L'  1978   7.9   
290022      290022        Péter    Rudolf               vegtigris  2001   8.7   
1900          1900       Amable   Aguiluz             www.XXX.com  2003   1.0   
198580      198580       Gianni     Lepre               ye for ye  1985   4.5   

              genre  
233670       Comedy  
216153  Documentary  
124273      Romance  
124272        Drama  
153855      Romance  
...             ...  
49586         Short  
165031        Short  
290022       Comedy  
1900        Romance  
198580        Drama  

[57422 rows x 7 columns]

'\nCount unique values from name column ...\n'

30247

'count of missing values in duplicates: '

Unnamed: 0    0
first_name    0
last_name     0
name          0
year          0
rank          0
genre         0
dtype: int64

Unnamed: 0    0.0
first_name    0.0
last_name     0.0
name          0.0
year          0.0
rank          0.0
genre         0.0
dtype: float64

Number of Rows in Aaltonen DataFrame : 11

Result dataframe :
     Unnamed: 0 first_name last_name                name  year  rank  \
78          78     Veikko  Aaltonen            Is meidn  1993   6.8   
79          79     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3   
80          80     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3   
81          81     Veikko  Aaltonen                 Maa  2001   6.0   
82          82     Veikko  Aaltonen          Merisairas  1996   4.3   
83          83     Veikko  Aaltonen          Merisairas  1996   4.3   
84          84     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9   
85          85     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9   
86          86     Veikko  Aaltonen           Tilinteko  1987   6.0   
87          87     Veikko  Aaltonen           Tilinteko  1987   6.0   
88          88     Veikko  Aaltonen       Tuhlaajapoika  1992   6.7   

          genre  
78        Drama  
79       Comedy  
80        Drama  
81  Document

Unnamed: 0 first_name last_name                                name  \
77              77  Rostislav     Aalto                        Cleaning Up!   
78              78     Veikko  Aaltonen                            Is meidn   
79              79     Veikko  Aaltonen                     Juoksuhaudantie   
80              80     Veikko  Aaltonen                     Juoksuhaudantie   
81              81     Veikko  Aaltonen                                 Maa   
...            ...        ...       ...                                 ...   
368837      368837     Memduh        Ün                               Garip   
368838      368838     Memduh        Ün                               Garip   
368861      368861     Memduh        Ün              Sezercik aslan parcasi   
368862      368862     Memduh        Ün              Sezercik aslan parcasi   
368891      368891       Idil      Üner  Liebenden vom Hotel von Osman, Die   

        year  rank        genre  
77      2002   8.3  Documentary  
78      1993   6.8        Drama  
79      2004   6.3       Comedy  
80      2004   6.3        Drama  
81      2001   6.0  Documentary  
...      ...   ...          ...  
368837  1986   7.9       Comedy  
368838  1986   7.9        Drama  
368861  1972   6.0        Drama  
368862  1972   6.0      Romance  
368891  2001   8.5        Short  

[113376 rows x 7 columns]

Number of Rows in Aaltonen DataFrame : 11

Result dataframe :
     Unnamed: 0 first_name last_name                name  year  rank  \
78          78     Veikko  Aaltonen            Is meidn  1993   6.8   
79          79     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3   
80          80     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3   
81          81     Veikko  Aaltonen                 Maa  2001   6.0   
82          82     Veikko  Aaltonen          Merisairas  1996   4.3   
83          83     Veikko  Aaltonen          Merisairas  1996   4.3   
84          84     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9   
85          85     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9   
86          86     Veikko  Aaltonen           Tilinteko  1987   6.0   
87          87     Veikko  Aaltonen           Tilinteko  1987   6.0   
88          88     Veikko  Aaltonen       Tuhlaajapoika  1992   6.7   

          genre  
78        Drama  
79       Comedy  
80        Drama  
81  Document

Unnamed: 0 first_name last_name                                name  \
77              77  Rostislav     Aalto                        Cleaning Up!   
78              78     Veikko  Aaltonen                            Is meidn   
79              79     Veikko  Aaltonen                     Juoksuhaudantie   
80              80     Veikko  Aaltonen                     Juoksuhaudantie   
81              81     Veikko  Aaltonen                                 Maa   
...            ...        ...       ...                                 ...   
368837      368837     Memduh        Ün                               Garip   
368838      368838     Memduh        Ün                               Garip   
368861      368861     Memduh        Ün              Sezercik aslan parcasi   
368862      368862     Memduh        Ün              Sezercik aslan parcasi   
368891      368891       Idil      Üner  Liebenden vom Hotel von Osman, Die   

        year  rank        genre  
77      2002   8.3  Documentary  
78      1993   6.8        Drama  
79      2004   6.3       Comedy  
80      2004   6.3        Drama  
81      2001   6.0  Documentary  
...      ...   ...          ...  
368837  1986   7.9       Comedy  
368838  1986   7.9        Drama  
368861  1972   6.0        Drama  
368862  1972   6.0      Romance  
368891  2001   8.5        Short  

[113376 rows x 7 columns]

In [76]:
from os import name
import pandas as pd

def misc_data(df):
  return


# program starts here
url = "https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"

movie_df = pd.read_csv(url)
df3 = movie_df[(movie_df['last_name'] == 'Aaltonen') & (movie_df['first_name'] == 'Veikko')]
display (df3)

# ================================================================
# remove all rows where column rank is null
# ================================================================
clean_df = df3[~df3["rank"].isnull()]
display (clean_df)

duplicates = clean_df[clean_df.duplicated(['name'],keep=False)]

display ("Duplicates: ", duplicates)


misc_data(movie_df)

Unnamed: 0 first_name last_name                name  year  rank  \
78          78     Veikko  Aaltonen            Is meidn  1993   6.8   
79          79     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3   
80          80     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3   
81          81     Veikko  Aaltonen                 Maa  2001   6.0   
82          82     Veikko  Aaltonen          Merisairas  1996   4.3   
83          83     Veikko  Aaltonen          Merisairas  1996   4.3   
84          84     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9   
85          85     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9   
86          86     Veikko  Aaltonen           Tilinteko  1987   6.0   
87          87     Veikko  Aaltonen           Tilinteko  1987   6.0   
88          88     Veikko  Aaltonen       Tuhlaajapoika  1992   6.7   
89          89     Veikko  Aaltonen         Tyvenluokka  2004   NaN   

          genre  
78        Drama  
79       Comedy  
80        Drama  
81  Documentary  
82        Drama  
83     Thriller  
84        Drama  
85     Thriller  
86        Crime  
87        Drama  
88     Thriller  
89  Documentary

Unnamed: 0 first_name last_name                name  year  rank  \
78          78     Veikko  Aaltonen            Is meidn  1993   6.8   
79          79     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3   
80          80     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3   
81          81     Veikko  Aaltonen                 Maa  2001   6.0   
82          82     Veikko  Aaltonen          Merisairas  1996   4.3   
83          83     Veikko  Aaltonen          Merisairas  1996   4.3   
84          84     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9   
85          85     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9   
86          86     Veikko  Aaltonen           Tilinteko  1987   6.0   
87          87     Veikko  Aaltonen           Tilinteko  1987   6.0   
88          88     Veikko  Aaltonen       Tuhlaajapoika  1992   6.7   

          genre  
78        Drama  
79       Comedy  
80        Drama  
81  Documentary  
82        Drama  
83     Thriller  
84        Drama  
85     Thriller  
86        Crime  
87        Drama  
88     Thriller

'Duplicates: '

Unnamed: 0 first_name last_name                name  year  rank     genre
79          79     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3    Comedy
80          80     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3     Drama
82          82     Veikko  Aaltonen          Merisairas  1996   4.3     Drama
83          83     Veikko  Aaltonen          Merisairas  1996   4.3  Thriller
84          84     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9     Drama
85          85     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9  Thriller
86          86     Veikko  Aaltonen           Tilinteko  1987   6.0     Crime
87          87     Veikko  Aaltonen           Tilinteko  1987   6.0     Drama



---


# **Cleaning the Dataset Notes**


---



1.  **Missing Values** - Column **rank** is the only col with missing values. It is a vital piece of info for our data analysis so I'm left with the choice somehow imputing a sensible value into this col or deleting all rows where it has null values or it is non-numeric. As there is no means to impute a sensible value, I have decided that I will need to disregard all rows where rank is not present in all further analysis.
2.   List item



### How many movies of each genre are there?
---

### Which director has the highest ranked movies?
---

### How many movies have ranks of over 9?
---


### Plot a bar chart of mean rank and genre
---

### Plot a pie chart of how many movies of each genre there are 
---


### Plot a graph showing the mean Rank for each year 

### What else can you find out from this dataset?
---
Make a plan of 3 further things you can do to interrogate and analyse this dataset

Type your answer here  

1.  
2.  
3.   

### Complete the tasks you have set out in the exercise above. 
---


# Reflection
----

## What skills have you demonstrated in completing this notebook?

Your answer:

## What caused you the most difficulty?

Your answer: